### PR TITLE
[rpc] Avoid wireDeserializer overreading buffers by 1 byte

### PR DESCRIPTION
**Pull request by jjlilley**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#36976 [rpc] Avoid wireDeserializer overreading buffers by 1 byte**

The bounds check and the read were swapped in two places - I noticed
ASAN complaining in an unrelated change on an erroneous buffer.

Adding a couple simple test cases.

Differential Revision: [D21148936](https://our.internmc.facebook.com/intern/diff/D21148936/)

### CODE REVIEW

File: test/cpp/rpc/test_wire_serialization.cpp [hunk 1]
```diff
@@ -68,6 +68,13 @@ TEST(WireSerialize, CloneSparseTensors) {
   EXPECT_TRUE(v3.get(0).is_same(sparse));
 }
 
+TEST(WireSerialize, Errors) {
+  EXPECT_THROW((void)torch::distributed::rpc::wireDeserialize("", 0), std::runtime_error);
+  EXPECT_THROW((void)torch::distributed::rpc::wireDeserialize(" ", 1), std::runtime_error);
```
Comment (mrshenli) [inline test/cpp/rpc/test_wire_serialization.cpp:73]: Shall we also check the error message here? "failed parse" vs "failed bounds"?

File: test/cpp/rpc/test_wire_serialization.cpp [hunk 2]
```diff
@@ -68,6 +68,13 @@ TEST(WireSerialize, CloneSparseTensors) {
...
+  auto serialized = torch::distributed::rpc::wireSerialize({}, {torch::randn({5, 5})});
+  EXPECT_THROW((void)torch::distributed::rpc::wireDeserialize(serialized.data(), serialized.size() / 2), std::runtime_error);
```
Comment (mrshenli) [inline test/cpp/rpc/test_wire_serialization.cpp:75]: this should throw "failed parse" as well?

**Review by mrshenli — APPROVED** [review]
Thanks for the fix!

File: test/cpp/rpc/test_wire_serialization.cpp (hunk 1 above)
Comment (jjlilley) [reply test/cpp/rpc/test_wire_serialization.cpp]: sure, just uploaded a version with this check

File: test/cpp/rpc/test_wire_serialization.cpp (hunk 2 above)
Comment (jjlilley) [reply test/cpp/rpc/test_wire_serialization.cpp]: This one is "failed bounds" actually - but thought that we should cover this somewhere

**Comment by jjlilley** [comment]
Thanks for looking, just uploaded a version with the extra unittest check

**Pull request closed without merging**